dnsnameresolver.go

```diff
@@ -0,0 +1,291 @@
+					// If details of DNS name and the DNSNameResolver objects already exist
+					// then check if the existing information match with the current one.
+					// Otherwise, don't proceed.
+					if dnsInfoMap[resolverObj.Namespace] != resolverObj.Name {
+						resolver.regularMapLock.Unlock()
+						return
+					}
+					delete(dnsInfoMap, resolverObj.Namespace)
+					if len(dnsInfoMap) > 0 {
+						resolver.regularDNSInfo[dnsName] = dnsInfoMap
+					} else {
+						delete(resolver.regularDNSInfo, dnsName)
+					}
+				}
+				resolver.regularMapLock.Unlock()
+			}
+
+			// Used only in unit tests.
+			if send != nil {
+				send(resolverObj)
+			}
+		},
+		// Used only in unit tests.
+		// Update event.
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// Get the DNSNameResolver object.
+			newResolverObj, ok := oldObj.(*dnsv1alpha1.DNSNameResolver)
+			if !ok {
+				log.Infof("object not of type DNSNameResolver: %v", oldObj)
+				return
+			}
+
+			// Check if namespace is configured or not.
+			if !resolver.configuredNamespace(newResolverObj.Namespace) {
+				return
+			}
+
+			// Used only in unit tests.
+			if send != nil {
+				send(newResolverObj)
+			}
+		},
+	})
+	return nil
+}
+
+// createNetworkClient returns a client supporting network.openshift.io apis.
+func createNetworkClient() (networkclient.Interface, error) {
+	kubeConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return networkclient.NewForConfig(kubeConfig)
+}
+
+// initPlugin initializes the ocp_dnsnameresolver plugin and returns the plugin startup and
+// shutdown callback functions.
+func (resolver *OCPDNSNameResolver) initPlugin() (func() error, func() error, error) {
+	err := resolver.initInformer(createNetworkClient, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resolver.stopCh = make(chan struct{})
+
+	onStart := func() error {
+		go func() {
+			resolver.dnsNameResolverInformer.Run(resolver.stopCh)
+		}()
+
+		timeout := 5 * time.Second
+		timeoutTicker := time.NewTicker(timeout)
+		defer timeoutTicker.Stop()
+		logDelay := 500 * time.Millisecond
+		logTicker := time.NewTicker(logDelay)
+		defer logTicker.Stop()
+		checkSyncTicker := time.NewTicker(100 * time.Millisecond)
+		defer checkSyncTicker.Stop()
+		for {
+			select {
+			case <-checkSyncTicker.C:
+				if resolver.dnsNameResolverInformer.HasSynced() {
+					return nil
+				}
+			case <-logTicker.C:
+				log.Info("waiting for DNS Name Resolver Informer sync before starting server")
+			case <-timeoutTicker.C:
+				log.Warning("starting server with unsynced DNS Name Resolver Informer")
+				return nil
+			}
+		}
+	}
+
+	onShut := func() error {
+		resolver.stopLock.Lock()
+		defer resolver.stopLock.Unlock()
+
+		// Only try draining the workqueue if we haven't already.
+		if !resolver.shutdown {
+			close(resolver.stopCh)
+			resolver.shutdown = true
+
+			return nil
+		}
+
+		return fmt.Errorf("shutdown already in progress")
```

Comment:
Why not just logging this? How CoreDNS will react to this error? Force stop?

Reply:
I followed a similar structure as the kubernetes coredns [plugin](https://github.com/openshift/coredns/blob/022a0530038602605b8f3e8866c2a6ded97708cc/plugin/kubernetes/controller.go#L317-L331). 

If any error is encountered while shutting down, then an exit code of 4 is returned: https://github.com/openshift/coredns/blob/022a0530038602605b8f3e8866c2a6ded97708cc/vendor/github.com/coredns/caddy/sigtrap.go#L67-L84.

Reply:
The assumption is that calling the shutdown for more than 1 time is an error, fair enough.